File: elixir/nth-prime/nth_prime.exs
defmodule Prime do

  @doc """
  Generates the nth prime.
  """
  @spec nth(non_neg_integer) :: non_neg_integer
  def nth(0), do: raise "0 is invalid!"
  def nth(count) do
    nth(count, 1)
  end
  defp nth(0, number), do: number
  defp nth(count, number) do
    number = number + 1

    if prime?(number) do
      nth(count - 1, number)
    else
      nth(count, number)
    end
  end

  defp prime?(number), do: prime?(number, top_factor_candiate(number))
  defp prime?(number, 1), do: true
  defp prime?(number, candidate) do
    if rem(number, candidate) === 0 do
      false
    else
      prime?(number, candidate - 1)
    end
  end

  defp top_factor_candiate(number) do
    number
    |> :math.sqrt
    |> Float.floor
    |> round
  end
end

File: ruby/nth-prime/nth_prime.rb
module BookKeeping
  VERSION = 1
end

class Prime
  def self.nth(count)
    raise ArgumentError.new("count must be at least 1!") unless count > 0

    last_prime = nil
    i = 1

    while count > 0
      i += 1
      next unless prime?(i)

      count -= 1
      last_prime = i
    end

    last_prime
  end

  # Check if a number is prime.
  # TODO: Make this fast enough for the big_prime test.
  def self.prime?(number)
    # special case for 2.
    return true if number == 2

    is_prime = true
    top_factor = Math.sqrt(number).floor

    # Check if any numbers below the number are a factor.
    (2..top_factor).each do |factor|
      if (number % factor).zero?
        is_prime = false
        break
      end
    end

    is_prime
  end
end
